### ISSUE TITLE
Pattern Matching for tuples in `let` isn't work as expected

### ISSUE THREAD

**Issue by astynax** [elm-lang/core#740]
Hello!
This code works fine (in `elm-0.17.1`):
```elm
import Html exposing (text)

main =
  let
    b = a / 2
    a = 4  -- declarations are in mixed order, but it's fine!
  in text <| toString (a, b)

-- Output:
-- (4, 2)
```
but this one isn't
```elm
main =
  let
    (a, b) = (4, a / 2)
  in text <| toString (a, b)

-- Output:
-- (4, NaN)
```
There is no warning / compilation errors. And I think what this binding is legal as legal the former one.

Here is an example with strings:
```elm
main =
  let (a, b) = ("hello", String.append a " world!")
  in text <| toString (a, b)

-- Output:
-- ("hello", "undefined world!")
```

**Comment by astynax** [comment]
- Browser: `Firefox 49.0`
- OS: `Ubuntu 16.04.1 LTS`
- Elm version: `0.17.1` (installed from `npm`)

**Comment by evancz** [comment]
This is a duplicate of https://github.com/elm-lang/elm-compiler/issues/873 which will be fixed in 0.18.

**Comment by lukewestby** [comment]
@astynax can you reopen this on `elm-compiler`? This is an issue with code generation, not with the core libraries.
